### PR TITLE
[#IOCIT-375] Added pubkey revoke and assertion deletion from redis to lock session function

### PR DESCRIPTION
**Pull request by michaeldisaro**
#### List of Changes

The `sessionLock` called by the User Data Delete Procedure has been updated to manage:
- the sending of the revoke pub key queue message 
- the `CF-assertionRef` binding deletion from redis

#### Motivation and Context

When a user asks for data deletion his lollipop public key MUST be revoked by sending a message to the revoke queue and his `CF-assertionRef` binding MUST be deleted from redis as we do with the session.

#### How Has This Been Tested?

- yarn build
- yarn test
- yarn lint

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/services/ISessionStorage.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export interface ISessionStorage {
    * @param user The AppUser value used to get the related fiscalCode
    */
```
Comment (Garma00) [inline src/services/ISessionStorage.ts:76]: We should update the doc as well

File: src/controllers/__tests__/sessionLockController.test.ts [hunk 2]
```diff
@@ -150,7 +207,80 @@ describe("SessionLockController#lockUserSession", () => {
...
+    const controller = new SessionLockController(
+      mockRedisSessionStorage,
+      mockRedisUserMetadataStorage,
+      mockLollipopService
+    );
+
+    const response = await controller.lockUserSession(req);
+    response.apply(res);
+
+    expect(res.status).toHaveBeenCalledWith(500);
```
Comment (Garma00) [inline src/controllers/__tests__/sessionLockController.test.ts:235]: We could add a check like 
```ts
expect(mockRevokePreviousAssertionRef).not.toHaveBeenCalled();
```

File: src/controllers/__tests__/sessionLockController.test.ts [hunk 3]
```diff
@@ -150,7 +207,80 @@ describe("SessionLockController#lockUserSession", () => {
...
```
Comment (Garma00) [inline src/controllers/__tests__/sessionLockController.test.ts:253]: Ditto, we could add a check like
```ts
expect(mockRevokePreviousAssertionRef).not.toHaveBeenCalled();
```

also if we create a mock for `delLollipopAssertionRefForUser` we could add a similar check
```ts
expect(delLollipopAssertionRefForUser).not.toHaveBeenCalled();
```

File: src/controllers/sessionLockController.ts [hunk 4]
```diff
@@ -96,6 +99,48 @@ export default class SessionLockController {
...
+              )
+            ),
+            // delete the assertionRef for the user
+            pipe(
+              TE.tryCatch(
+                () =>
+                  this.sessionStorage.delLollipopAssertionRefForUser(
+                    fiscalCode
+                  ),
+                E.toError
```
Comment (Garma00) [inline src/controllers/sessionLockController.ts:140]: Even if `getLollipopAssertionRefForUser` or `revokePreviousAssertionRef` returns left in the previous pipe we are calling `delLollipopAssertionRefForUser` here, are we ok with that?

Comment (michaeldisaro) [reply src/controllers/sessionLockController.ts:140]: It is an applicative sequence. It should break if something goes left.

File: src/controllers/__tests__/sessionLockController.test.ts (hunk 2 above)
Comment (michaeldisaro) [reply src/controllers/__tests__/sessionLockController.test.ts]: done https://github.com/pagopa/io-backend/pull/995/commits/1216de4e971ce08cf7c7b7b5f9939dfdc4845153

File: src/services/ISessionStorage.ts (hunk 1 above)
Comment (michaeldisaro) [reply src/services/ISessionStorage.ts]: done https://github.com/pagopa/io-backend/pull/995/commits/1216de4e971ce08cf7c7b7b5f9939dfdc4845153

File: src/controllers/sessionLockController.ts [hunk 5]
```diff
@@ -96,6 +99,48 @@ export default class SessionLockController {
               ),
               TE.chain(TE.fromEither)
             ),
+            // revoke pubkey
+            pipe(
+              TE.tryCatch(
+                () =>
+                  // retrieve the assertionRef for the user
+                  this.sessionStorage.getLollipopAssertionRefForUser(
```
Comment (BurnedMarshal) [inline src/controllers/sessionLockController.ts:107]: ditto

File: src/controllers/authenticationController.ts [hunk 6]
```diff
@@ -477,7 +479,9 @@ export default class AuthenticationController {
                 TE.tryCatch(
                   () =>
                     // retrieve the assertionRef for the user
-                    this.sessionStorage.getLollipopAssertionRefForUser(user),
+                    this.sessionStorage.getLollipopAssertionRefForUser(
```
[the remark this answers is not in the document]
Comment (michaeldisaro) [reply src/controllers/authenticationController.ts:482]: This feature is not available until we will update to Redis 6.2.0. I tracked a Jira chore issue about this.

File: src/controllers/__tests__/sessionLockController.test.ts [hunk 7]
```diff
@@ -150,12 +213,97 @@ describe("SessionLockController#lockUserSession", () => {
...
+
+    const controller = new SessionLockController(
+      mockRedisSessionStorage,
+      mockRedisUserMetadataStorage,
+      mockLollipopService
+    );
+
+    const response = await controller.lockUserSession(req);
+    response.apply(res);
+
```
Comment (arcogabbo) [inline src/controllers/__tests__/sessionLockController.test.ts:284]:
```suggestion
    expect(mockRevokePreviousAssertionRef).toHaveBeenCalled();
```

File: src/controllers/__tests__/sessionLockController.test.ts [hunk 8]
```diff
@@ -150,12 +213,97 @@ describe("SessionLockController#lockUserSession", () => {
...
+    });
+
+    const controller = new SessionLockController(
+      mockRedisSessionStorage,
+      mockRedisUserMetadataStorage,
+      mockLollipopService
     );
 
     const response = await controller.lockUserSession(req);
     response.apply(res);
```
[the remark this answers is not in the document]
Comment (michaeldisaro) [reply src/controllers/__tests__/sessionLockController.test.ts]: done https://github.com/pagopa/io-backend/pull/995/commits/979fdf8e9405d3fa923e7fe01d3794936f87d3ec

**Review by BurnedMarshal — APPROVED** [review]

**Pull request merged**